### PR TITLE
Transition Animations

### PR DESCRIPTION
**Pull request by dantheli**
Adds fun transitions to most VC presentations involving series and episodes

### CODE REVIEW

File: Podcast/AppDelegate.swift [hunk 1]
```diff
@@ -64,6 +65,9 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         } catch {
             print("No AudioSession!! Don't know what do to here. ")
         }
+
+        // Hero
+        Hero.shared.containerColor = .white
```
Comment (natashaarmbrust) [inline Podcast/AppDelegate.swift]: .offWhite

File: Podcast/BookmarkViewController.swift [hunk 2]
```diff
@@ -50,6 +50,14 @@ class BookmarkViewController: ViewController, EmptyStateTableViewDelegate, UITab
             }
         }
     }
+
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+
+        navigationController?.isHeroEnabled = true
+        navigationController?.heroNavigationAnimationType = .selectBy(presenting: .zoom, dismissing: .zoomOut)
+        navigationController?.interactivePopGestureRecognizer?.delegate = nil
```
Comment (natashaarmbrust) [inline Podcast/BookmarkViewController.swift:59]: why do we need to assign the delegate to nil .. won't it automatically be nil

File: Podcast/DiscoverViewController.swift [hunk 3]
```diff
@@ -130,7 +130,8 @@ class DiscoverViewController: ViewController, UITableViewDelegate, UITableViewDa
     
     func recommendedSeriesTableViewCell(cell: RecommendedSeriesTableViewCell, didSelectItemAt indexPath: IndexPath) {
         
-        let seriesDetailViewController = SeriesDetailViewController(series: series[indexPath.row])
+        let seriesDetailViewController = SeriesDetailViewController()
+        seriesDetailViewController.series = series[indexPath.row]
```
Comment (natashaarmbrust) [inline Podcast/DiscoverViewController.swift]: I am getting a crash now when doing to a series detail from profile which is the result of changing this code

File: Podcast/EpisodeDetailViewController.swift [hunk 4]
```diff
@@ -41,18 +43,25 @@ class EpisodeDetailViewController: ViewController, EpisodeDetailHeaderViewDelega
...
+        let style = NSMutableParagraphStyle()
+        let attributedDescription = NSMutableAttributedString(attributedString: episode.attributedDescription)
+        attributedDescription.addAttribute(NSAttributedStringKey.paragraphStyle, value: style, range: NSMakeRange(0, attributedDescription.length))
+        episodeDescriptionView.attributedText = attributedDescription
+        // weird known iOS bug when resizing a textContainer's text to be the start of a UITextView .. do not remove
+        episodeDescriptionView.isScrollEnabled = false
+        episodeDescriptionView.setNeedsUpdateConstraints()
+        episodeDescriptionView.isScrollEnabled = true
+
+        isHeroEnabled = true
```
Comment (natashaarmbrust) [inline Podcast/EpisodeDetailViewController.swift:57]: can we slow down the velocity of the image movement?

File: Podcast/FeedViewController.swift [hunk 5]
```diff
@@ -58,6 +59,10 @@ class FeedViewController: ViewController {
         feedTableView.refreshControl?.addTarget(self, action: #selector(fetchCards), for: .valueChanged)
         
         fetchCards()
+
+        navigationController?.view.backgroundColor = .white
```
Comment (natashaarmbrust) [inline Podcast/FeedViewController.swift]: .offWhite!

File: Podcast/FeedViewController.swift [hunk 6]
```diff
@@ -58,6 +59,10 @@ class FeedViewController: ViewController {
...
+        navigationController?.heroNavigationAnimationType = .selectBy(presenting: .zoom, dismissing: .zoomOut)
+        navigationController?.interactivePopGestureRecognizer?.delegate = nil
```
Comment (natashaarmbrust) [inline Podcast/FeedViewController.swift]: again with the nil

File: Podcast/FeedViewController.swift [hunk 7]
```diff
@@ -140,9 +151,22 @@ extension FeedViewController: FeedElementTableViewCellDelegate, EmptyStateTableV
...
             viewController.episode = episode
+
+            let cell = tableView.cellForRow(at: indexPath) as? FeedElementTableViewCell
+            if let subjectView = cell?.subjectView as? EpisodeSubjectView {
+                viewController.placeholderImage = subjectView.podcastImage.image
+            }
+
             navigationController?.pushViewController(viewController, animated: true)
         case .followingSubscription(_, let series):
-            let viewController = SeriesDetailViewController(series: series)
```
Comment (natashaarmbrust) [inline Podcast/FeedViewController.swift]: again why are we separating these?

File: Podcast/SeriesDetailViewController.swift [hunk 8]
```diff
@@ -17,30 +17,32 @@ class SeriesDetailViewController: ViewController, SeriesDetailHeaderViewDelegate
...
     var continueInfiniteScroll = true
     var currentlyPlayingIndexPath: IndexPath?
+
+    var viewAppeared = false
+    var episodesFetched = false
+    var animated = false
     
     var episodes: [Episode] = []
-    
-    convenience init(series: Series) {
```
Comment (natashaarmbrust) [inline Podcast/SeriesDetailViewController.swift]: whyyy?

File: Podcast/SeriesDetailViewController.swift [hunk 9]
```diff
@@ -79,19 +81,30 @@ class SeriesDetailViewController: ViewController, SeriesDetailHeaderViewDelegate
         }
         
         loadingAnimation.startAnimating()
-        
-        if let series = series {
-            setSeries(series: series)
-        }
+
+        setSeries(series: series)
```
Comment (natashaarmbrust) [inline Podcast/SeriesDetailViewController.swift]: I get a crash on the setSeries line

File: Podcast/SubscriptionsViewController.swift [hunk 10]
```diff
@@ -56,7 +61,12 @@ class SubscriptionsViewController: ViewController, UICollectionViewDelegate, UIC
     }
     
     func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
-        let seriesDetailViewController = SeriesDetailViewController(series: subscriptions[indexPath.row])
+        let seriesDetailViewController = SeriesDetailViewController()
```
Comment (natashaarmbrust) [inline Podcast/SubscriptionsViewController.swift]: fix

**Review by natashaarmbrust — CHANGES_REQUESTED** [review]

**Comment by natashaarmbrust** [comment]
Add to listening history as well

**Comment by natashaarmbrust** [comment]
Try to add to search

File: Podcast/DiscoverViewController.swift (hunk 3 above)
Comment (dantheli) [reply Podcast/DiscoverViewController.swift]: fixed

File: Podcast/SeriesDetailViewController.swift (hunk 9 above)
Comment (dantheli) [reply Podcast/SeriesDetailViewController.swift]: fixed

File: Podcast/BookmarkViewController.swift (hunk 2 above)
Comment (dantheli) [reply Podcast/BookmarkViewController.swift:59]: It is removed when you do custom transitions

File: Podcast/FeedViewController.swift [hunk 11]
```diff
@@ -58,11 +58,14 @@ class FeedViewController: ViewController {
         feedTableView.refreshControl?.addTarget(self, action: #selector(fetchCards), for: .valueChanged)
         
         fetchCards()
+
+        navigationController?.heroNavigationAnimationType = .selectBy(presenting: .zoom, dismissing: .zoomOut)
```
Comment (mindylou) [inline Podcast/FeedViewController.swift:62]: i'm getting this weird glitch on returning to feed from series detail

File: Podcast/FeedViewController.swift [hunk 12]
```diff
@@ -100,18 +109,22 @@ class FeedViewController: ViewController {
...
             self.feedTableView.finishInfiniteScroll()
-            self.feedTableView.reloadData()
+            self.reloadTableView()
         }
         
         System.endpointRequestQueue.addOperation(fetchFeedEndpointRequest)
     }
+
+    func reloadTableView() {
+        feedTableView.reloadSections(IndexSet(integer: 0), with: .fade)
```
Comment (mindylou) [inline Podcast/FeedViewController.swift:125]: not sure if the fade animation should be applied every time the view appears (such as every time you tap on the tab bar for feed), it's kind of distracting

**Comment by natashaarmbrust** [comment]
Some of the transitions are a bit funky ... Really try to play with the app on the simulator on multiple phone types to fix some of these bugs. Also I am still getting a crash when going from an `EpisodeDetailViewController` to a`SeriesDetailViewController` via the title button.

File: Podcast/FeedViewController.swift (hunk 11 above)
Comment (dantheli) [reply Podcast/FeedViewController.swift:62]: Will fix by not refreshing the feed every time we look at it

File: Podcast/FeedViewController.swift (hunk 12 above)
Comment (dantheli) [reply Podcast/FeedViewController.swift:125]: See above

**Pull request closed without merging**